src/vehicle.cpp

```diff
@@ -5632,9 +5635,15 @@ int vehicle::damage_direct( int p, int dmg, damage_type type )
 
     if( parts[p].is_tank() ) {
         explode_fuel( p, type );
-    } else if( parts[ p ].is_broken() && part_flag(p, "UNMOUNT_ON_DAMAGE") ) {
-        g->m.spawn_item( global_part_pos3( p ), part_info( p ).item, 1, 0, calendar::turn );
-        remove_part( p );
+    }
+
+    if( parts[p].is_broken() ) {
+        if( part_flag(p, "UNMOUNT_ON_DAMAGE") ) {
+            g->m.spawn_item( global_part_pos3( p ), part_info( p ).item, 1, 0, calendar::turn );
+            remove_part( p );
+        }
+
+        refresh();
```

Comment:
Forcing `refresh` here to keep that weird thing with `floating` not having broken parts sounds really wrong.
Instead make a `can_float` or something like that, which checks if there is a non-broken floating part or something like that.